[redditr] Fix extraction of i.redd.it animated GIFs

jwilk opened the pull request:
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

When a Reddit post embedded an [i.redd.it](https://i.redd.it/) animated GIF, youtube-dl would download the raw GIF, even though better format (MP4) was available.

Before:
```console
$ youtube-dl https://www.reddit.com/r/WatchPeopleDieInside/comments/7kb915/first_of_all/
[RedditR] 7kb915: Downloading JSON metadata
[generic] aspakv7pfvry: Requesting header
WARNING: Falling back on generic information extractor.
[generic] aspakv7pfvry: Downloading webpage
WARNING: URL could be a direct video link, returning it as such.
[download] Destination: First of all...-aspakv7pfvry.gif
[download] 100% of 17.72MiB in 00:00
```

After:
```console
$ youtube-dl https://www.reddit.com/r/WatchPeopleDieInside/comments/7kb915/first_of_all/
[RedditR] 7kb915: Downloading JSON metadata
[download] Destination: First of all...-7kb915.mp4
[download] 100% of 945.60KiB in 00:00
```

jwilk commented:
Ho-hum, 022d336e9213f17846838b9a9bb556d8f0b1a846 doesn't work for some links, for example:
https://www.reddit.com/r/ProgrammerHumor/comments/7kk22g/what_working_with_css_feels_like/

The anti-gif-bot somehow got its hands of the MP4:
https://www.reddit.com/r/ProgrammerHumor/comments/7kk22g/what_working_with_css_feels_like/dreyplg/
No sure how it managed to do that. :-/

dstftw commented on youtube_dl/extractor/reddit.py:
```diff
@@ -99,6 +118,24 @@ def _real_extract(self, url):
+            variants = data['preview']['images'][0]['variants']
+            FORMATS = ('gif', 'mp4')
+            quality = qualities(FORMATS)
+            for fmt in FORMATS:
```
Iteration over dict not the hardcoded list.

dstftw commented on youtube_dl/extractor/reddit.py:
```diff
@@ -99,6 +118,24 @@ def _real_extract(self, url):
+            quality = qualities(FORMATS)
+            for fmt in FORMATS:
+                fmt_variants = variants[fmt]
+                for variant in [fmt_variants['source']] + fmt_variants['resolutions']:
```
Should be tolerate to missing keys.

dstftw commented on youtube_dl/extractor/reddit.py:
```diff
@@ -99,6 +118,24 @@ def _real_extract(self, url):
+                fmt_variants = variants[fmt]
+                for variant in [fmt_variants['source']] + fmt_variants['resolutions']:
+                    formats.append({
+                        'url': unescapeHTML(variant['url']),
```
Same.

dstftw commented on youtube_dl/extractor/reddit.py:
```diff
@@ -99,6 +118,24 @@ def _real_extract(self, url):
+                        'ext': fmt,
+                        'quality': quality(fmt),
+                        'width': variant.get('width'),
+                        'height': variant.get('height'),
```
`int_or_none`.

jwilk commented:
> Ho-hum, 022d336e9213f17846838b9a9bb556d8f0b1a846 doesn't work for some links, for example:
> https://www.reddit.com/r/ProgrammerHumor/comments/7kk22g/what_working_with_css_feels_like/

I still don't understand what's going on, but I implemented fallback so that at least you get *something* when this happens.
I also fixed the problems pointed out in the review.

jwilk commented:
Any news on this?
